### ISSUE TITLE
display-name acceptTranspilerName should not accept anonymous functional components

### ISSUE THREAD

**Issue by epeli** [yannickcr/eslint-plugin-react#367]
When anonymous function are used the component does not get a name which make debugging harder because error messages are less useful:

    Warning: Failed propType: Required prop `height` was not specified in `Component`. Check the render method of `Component`.

So the these would be problems:

```js
const Foo = ({children}) => <div>{children}</div>;

var Bar = function(props) {
    return <div>{props.children}</div>;
};
```

and these would not be

```js
const Foo = ({children}) => <div>{children}</div>;
Foo.name = "Foo";

function Bar(props) {
    return <div>{props.children}</div>;
};
```

**Comment by yannickcr** [comment]
That's the purpose of the `display-name` rule, especially with the `acceptTranspilerName` option that should tell to the rule to only force you to set a `displayName` if the transpiler cannot find a name for the component.

But it seems it does not works the way I thought for functional components.
```js
const Foo = (props) => <div>Hello {name}</div>;
Foo.propTypes = {
  name: React.PropTypes.string.isRequired
};
```
Will trigger:
```
Warning: Failed propType: Required prop `hello` was not specified in `<<anonymous>>`.
```
And for a classic comment (in ES5 or ES6):
```js
const Foo = React.createClass({
  propTypes: {
    hello: React.PropTypes.string.isRequired
  },
  render: function() {
    return <div>{this.props.hello}</div>;
  }
});
```
Will trigger:
```
Warning: Failed propType: Required prop `hello` was not specified in `Foo`.
```
So it's a bug in the `acceptTranspilerName` option.

**Comment by epeli** [comment]
Ah, you're right this should be in the `display-name` rule. But I'm not quite sure how it relates to `acceptTranspilerName` because at least Babel does not add a transpile name for function components?

I think this should always give a warning:

```js
const Hello = () => <div>Hello</div>;
```

which can be fixed by manually assigning a name:

```js
Hello.name = "Hello";
```

or by using a function declaration.

**Comment by yannickcr** [comment]
> But I'm not quite sure how it relates to `acceptTranspilerName` because at least Babel does not add a transpile name for function components?

I thought he adds one, but obviously it is not the case. So you're right
```js
const Hello = () => <div>Hello</div>;
```
should warn.

**Comment by epeli** [comment]
I was so wrong about this. In ES2015 `.name` is actually picked from the variable name automatically and Babel ES2015 preset implements it correctly.

http://www.2ality.com/2015/09/function-names-es6.html
